stan/math/prim/mat/matrix_addons.h

```diff
@@ -0,0 +1,203 @@
+/**
+ * Structure to determine whether input struct has member named "d_".
+ * This is used to differentiate between var and fvar<T> without
+ * depending on external code.
+ *
+ * Re-implements std::void_t for pre-c++17 to detect ill-formed types
+ * in SFINAE
+ */
+template<class, class = void>
+struct Hasd_ : std::false_type
+{ };
+template<class T>
+struct Hasd_<T, decltype((void)(T::d_))> : std::true_type
+{ };
+
+/**
+ * Structure to return a view to the values in a var, vari*, and fvar<T>.
+ * To identify the correct member to call for a given input, templates
+ * check a combination of whether the input is a pointer (i.e. vari*)
+ * and/or whether the input has member ".d_" (i.e. fvar).
+ *
+ * The operators are overloaded for both const and non-const inputs
+ */
+struct val_Op{
+  EIGEN_EMPTY_STRUCT_CTOR(val_Op);
```

Comment:
Some some Eigen specifics it may be nice to doc them with a link to the place in Eigen where they are doc'd

https://eigen.tuxfamily.org/dox/XprHelper_8h_source.html